Essentials/src/com/earth2me/essentials/Teleport.java

```diff
@@ -121,28 +121,38 @@ public void now(Player entity, boolean cooldown, TeleportCause cause) throws Exc
     protected void now(IUser teleportee, ITarget target, TeleportCause cause) throws Exception {
         cancel(false);
         teleportee.setLastLocation();
-        Location loc = target.getLocation();
 
-        if (LocationUtil.isBlockUnsafeForUser(teleportee, loc.getWorld(), loc.getBlockX(), loc.getBlockY(), loc.getBlockZ())) {
-            if (ess.getSettings().isTeleportSafetyEnabled()) {
-                if (ess.getSettings().isForceDisableTeleportSafety()) {
-                    PaperLib.teleportAsync(teleportee.getBase(), loc, cause);
+        PaperLib.getChunkAtAsync(target.getLocation()).thenAccept(chunk -> {
+            Location loc = target.getLocation();
+            if (LocationUtil.isBlockUnsafeForUser(teleportee, chunk, loc.getBlockX(), loc.getBlockY(), loc.getBlockZ())) {
+                if (ess.getSettings().isTeleportSafetyEnabled()) {
+                    if (ess.getSettings().isForceDisableTeleportSafety()) {
+                        PaperLib.teleportAsync(teleportee.getBase(), loc, cause);
+                    } else {
+                        try {
+                            PaperLib.teleportAsync(teleportee.getBase(), LocationUtil.getSafeDestination(ess, teleportee, loc), cause);
+                        } catch (Exception e) {
+                            e.printStackTrace();
+                        }
+                    }
                 } else {
-                    PaperLib.teleportAsync(teleportee.getBase(), LocationUtil.getSafeDestination(ess, teleportee, loc), cause);
+                    try {
+                        throw new Exception(tl("unsafeTeleportDestination", loc.getWorld().getName(), loc.getBlockX(), loc.getBlockY(), loc.getBlockZ()));
+                    } catch (Exception e) {
+                        e.printStackTrace();
+                    }
```

Comment:
I think you have hit the crux of why this hasn't happened previously :)

We can't throw the exception back upstream obviously because we are in a different stack frame. e.printStackTrace() isn't remotely acceptable, and even if it was, we lost the entire point of the exception - to notify the user that the destination of the teleport is unsafe. This is bad design in general (using exceptions as user feedback mechanisms), but unfortunately unless we are going to uproot the entire codebase its here to stay.

A possible solution is to give now() a separate argument which accepts an exception as a callback, but that would bubble all the way up the stack as Node-esque hell. Not sure what the right solution here is.

Reply:
I looked into it, creating a callable is **way** too much of hassle to do here than just printing the stack trace. The point of these exceptions are to inform the user of their own mistake. As you pointed out, these are poor uses of exceptions and would require an uproot of the entire codebase to get rid of these poor uses. That being said, I feel like that is not in the scope of this PR and should be addressed separately than hacking in something that would only be used here and not anywhere else.

Reply:
I would like the feedback of other team members on this as well!

CC: @pop4959 @md678685

Reply:
Unfortunately without somehow signalling up to the user how things broke, this is a regression of behavior. I would also agree that fixing that is not in the scope of this PR, but breaking user experience changes like this has been something that EssentialsX has always tried to avoid (to its own peril a lot of the time sure, but then again server owners are an interesting bunch).

https://xkcd.com/1172/

Reply:
I haven't had the time to properly look into or test this PR yet, however I'll just say that existing behavior should be retained, including providing correct errors back to the user (and wherever else). If it's not possible to work around how Essentials' exception system works then it definitely blocks this PR and that will need to be taken care of first (perhaps by doing some refactoring so that the user errors can be accessed without needing to go through an exception). As Roy mentioned it is not in scope for this PR but it would certainly need to be addressed before pulling this.

MD has been somewhat busy lately, so it may be a while for him to give some assessment on this, but I'll try to give this PR more love soon. We should definitely discuss this carefully and not rush to push out these changes before they are completely ready.